File: java/09_pgm_avion_interface/src/co/edu/campusucc/poo/implement/TipoCarga.java
package co.edu.campusucc.poo.implement;

import co.edu.campusucc.poo.abst.Avion;
import co.edu.campusucc.poo.interfaces.IAvionCarga;

public class TipoCarga extends Avion implements IAvionCarga{

    public TipoCarga(){
        super();
        setTipo("Carga");
    }
    
    @Override
    public void startingEngine() {
        System.out.println("⛽Check Fuel ✅");
        for (int i = 0; i < 3; i++) {
            System.out.println("💥💥Spark💥💥");
            try {
                Thread.sleep(800 - (i * 400));
            } catch (InterruptedException e) {
                System.err.println("⛔:" + e);
            }
        }
        System.out.println("Started the Motor...✈️✈️");
    }

    @Override
    public void speedUp() {
        String speedUpString = "speed up in Progress: ";
        for (int i = 0; i < super.getVelocidad(); i++) {
            speedUpString += "✈️";
        }

        for (int i = 0; i < 10; i++) {
            speedUpString += "✈️";
            System.out.println(speedUpString);
            try {
                Thread.sleep(500 - (i * 50));
            } catch (InterruptedException e) {
                System.err.println("⛔:" + e);
            }
        }
        super.addSpeed(10);
    }

    @Override
    public void stopingEngine() {
        System.out.println("✅ Check the Turbine 🆗🆗");
        for (int i = 0; i < 2; i++) {
            System.out.println("⛔ turning off turbine❌❌");
            try {
                Thread.sleep(800 - (i * 400));
            } catch (InterruptedException e) {
                System.err.println("⛔:" + e);
            }
        }
        System.out.println("Stoped...⚡⚡");
    }
    
    @Override
    public void openDoors() {
        System.out.println("✅ Open Door 🆗");
    }
    @Override
    public void closeDoors() {
        System.out.println("✅ Close Door 🆗");
    }
    @Override
    public void pickUp() {
        System.out.println("✅ PickUp 🆗");
    }
    
    @Override
    public void chargeFuel(int litros) {
        super.chargeFuel(litros);
        System.out.println("");
        for (int i = 0; i < litros; i++) {
            try {
                System.out.print("⛽");
                Thread.sleep(100);
            } catch (InterruptedException e) {
                System.err.println("⛔:" + e);
            }
        }
        System.out.println("");
        System.out.println("⛽ Full Fuel ✅");
    }
    
}

File: c++/09_pgm_avion_interface/src/co/edu/campusucc/poo/implement/TipoCarga.cpp
#include <iostream>
#include <string>
#include "Avion.h"
#include "IAvionCarga.h"
using namespace std;

class TipoCarga : public Avion, public IAvionCarga {
public:
    TipoCarga() : Avion() {
        setTipo("Carga");
    }

    void startingEngine() override {
        cout << "⛽Check Fuel ✅" << endl;
        for (int i = 0; i < 3; i++) {
            cout << "💥💥Spark💥💥" << endl;
            try {
                std::this_thread::sleep_for(std::chrono::milliseconds(800 - (i * 400)));
            } catch (std::exception& e) {
                cerr << "⛔:" << e.what() << endl;
            }
        }
        cout << "Started the Motor...✈️✈️" << endl;
    }

    void speedUp() override {
        string speedUpString = "speed up in Progress: ";
        for (int i = 0; i < getVelocidad(); i++) {
            speedUpString += "✈️";
        }

        for (int i = 0; i < 10; i++) {
            speedUpString += "✈️";
            cout << speedUpString << endl;
            try {
                std::this_thread::sleep_for(std::chrono::milliseconds(500 - (i * 50)));
            } catch (std::exception& e) {
                cerr << "⛔:" << e.what() << endl;
            }
        }
        addSpeed(10);
    }

    void stopingEngine() override {
        cout << "✅ Check the Turbine 🆗🆗" << endl;
        for (int i = 0; i < 2; i++) {
            cout << "⛔ turning off turbine❌❌" << endl;
            try {
                std::this_thread::sleep_for(std::chrono::milliseconds(800 - (i * 400)));
            } catch (std::exception& e) {
                cerr << "⛔:" << e.what() << endl;
            }
        }
        cout << "Stoped...⚡⚡" << endl;
    }

    void openDoors() override {
        cout << "✅ Open Door 🆗" << endl;
    }

    void closeDoors() override {
        cout << "✅ Close Door 🆗" << endl;
    }

    void pickUp() override {
        cout << "✅ PickUp 🆗" << endl;
    }

    void chargeFuel(int litros) override {
        Avion::chargeFuel(litros);
        cout << endl;
        for (int i = 0; i < litros; i++) {
            try {
                cout << "⛽";
                std::this_thread::sleep_for(std::chrono::milliseconds(100));
            } catch (std::exception& e) {
                cerr << "⛔:" << e.what() << endl;
            }
        }
        cout << endl;
        cout << "⛽ Full Fuel ✅" << endl;
    }
};
